### PR TITLE
Fix subresource update logic.

### PR DESCRIPTION
**Pull request by vaikas**
Signed-off-by: Ville Aikas <vaikas@chainguard.dev>

# Changes
While working on: https://github.com/sigstore/policy-controller/pull/88 I noticed that subresources are not being correctly handled.

🐛 Fix a bug where subresource was not correctly set when doing webhook validation.

-
-
-

/kind bug

Fixes #

**Release Note**

```release-note
Fix a bug where subresource updates were being ignored.
```

**Docs**

```docs

```

### CODE REVIEW

File: apis/contexts.go [hunk 1]
```diff
@@ -85,14 +85,21 @@ func IsInUpdate(ctx context.Context) bool {
 	return ctx.Value(inUpdateKey{}) != nil
 }
 
-// IsInStatusUpdate checks whether the context is an Update.
-func IsInStatusUpdate(ctx context.Context) bool {
+// IsInSubResourceUpdate returns the subresource being updated or "" if there
+// is no subresource that's being updated. Examples are "status" for Status
+// updates, or "scale" for scaling Deployment.
+func IsInSubResourceUpdate(ctx context.Context) string {
```
Comment (mattmoor) [inline apis/contexts.go:91]: Calling is `Is...` feels weird returning a string.  Maybe `GetUpdatedSubresource`?

File: webhook/resourcesemantics/validation/validation_admit.go [hunk 2]
```diff
@@ -141,13 +141,13 @@ func (ac *reconciler) decodeRequestAndPrepareContext(
...
 
-	if newObj != nil && oldObj != nil && req.SubResource == "" {
-		ctx = apis.WithinSubResourceUpdate(ctx, oldObj, req.SubResource)
-	}
-
 	switch req.Operation {
 	case admissionv1.Update:
-		ctx = apis.WithinUpdate(ctx, oldObj)
+		if req.SubResource != "" {
+			ctx = apis.WithinSubResourceUpdate(ctx, oldObj, req.SubResource)
```
Comment (mattmoor) [inline webhook/resourcesemantics/validation/validation_admit.go:147]: Hmm, does this return the right thing for `apis.IsInUpdate`?

**Review by mattmoor — COMMENTED** [review]
Couple nits, but generally LGTM.

thanks for the bug fix.

File: webhook/resourcesemantics/validation/validation_admit.go (hunk 2 above)
Comment (vaikas) [reply webhook/resourcesemantics/validation/validation_admit.go:147]: yes:
https://github.com/knative/pkg/pull/2546/files/3e662c0cea7c64938e64f45d5c46fef3b4c5741a

**Comment by dprotaso** [comment]
fixes: https://github.com/knative/pkg/issues/2184

/lgtm
/approve

File: webhook/resourcesemantics/validation/validation_admit.go [hunk 3]
```diff
@@ -141,13 +141,13 @@ func (ac *reconciler) decodeRequestAndPrepareContext(
...
+		} else {
+			ctx = apis.WithinUpdate(ctx, oldObj)
+		}
```
[the remark this answers is not in the document]
Comment (vaikas) [reply webhook/resourcesemantics/validation/validation_admit.go:150]: ok, phew :) Yeah, so I also checked the defaulting behaviour and it's precisely like this. This was just b0rk3d here.
https://github.com/knative/pkg/blob/main/webhook/resourcesemantics/defaulting/defaulting.go#L422

**Pull request merged**